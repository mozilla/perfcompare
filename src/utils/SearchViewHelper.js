import store from '../common/store';
import {
  updateSearchValue,
  updateSearchResults,
  updateRepository,
} from '../reducers/searchSlice';
import {
  fetchRecentRevisions,
  fetchRevisionByID,
  fetchRevisionsByAuthor,
} from '../thunks/searchThunk';

const SearchViewHelper = {
  searchByRevisionOrEmail(repository, search) {
    const emailMatch = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    const longHashMatch = /[a-zA-Z0-9]{40}/;

    if (emailMatch.test(search)) {
      store.dispatch(fetchRevisionsByAuthor({ repository, search }));
    } else if (longHashMatch.test(search)) {
      store.dispatch(fetchRevisionByID({ repository, search }));
    }
  },

  handleChangeDropdown(event) {
    const repository = event.target.innerText;
    const search = store.getState().search.searchValue;

    // Update state with selected repository
    store.dispatch(updateRepository(repository));
    // If repository is selected after search value is input, fetch search results
    if (search !== '') {
      SearchViewHelper.searchByRevisionOrEmail(repository, search);
    }
    // Fetch 10 most recent revisions when repository changes
    store.dispatch(fetchRecentRevisions(repository));
  },

  handleChangeSearch(event) {
    const search = event.target.value;
    const { repository } = store.getState().search;
    store.dispatch(updateSearchValue(search));

    // If search input is cleared, clear results
    if (search === '') store.dispatch(updateSearchResults([]));

    if (repository === '') {
      console.log('Please select a repository');
    } else {
      SearchViewHelper.searchByRevisionOrEmail(repository, search);
    }
  },
};

export default SearchViewHelper;
