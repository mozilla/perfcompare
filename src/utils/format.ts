const dateFormatter = new Intl.DateTimeFormat('en-US', {
  dateStyle: 'medium',
});

export function formatDate(date: Date) {
  return dateFormatter.format(date);
}

export function formatDateRange(date1: Date, date2: Date) {
  return formatDate(date1) + ' - ' + formatDate(date2);
}
