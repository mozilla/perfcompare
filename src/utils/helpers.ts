import {
  frameworkMap,
  devToolsFramework,
  baseDocsURL,
  removedOldTestDevTools,
  nonDocumentedTestsDevTools,
  supportedPerfdocsFrameworks,
} from '../common/constants';
import type { Repository, Changeset } from '../types/state';
import type { Framework, SupportedPerfdocsFramework } from '../types/types';

const truncateHash = (revision: Changeset['revision']) => revision.slice(0, 12);

// return only most recent commit message
// first commit is usually 'try_task_config'
const getLatestCommitMessage = (item: Changeset) => {
  const { repository_id: repositoryId, revisions } = item;
  const isTry = repositoryId === 4;
  const lastUsefulRevision =
    isTry && revisions.length > 1 ? revisions[1] : revisions[0];
  const lastUsefulSummary = lastUsefulRevision.comments.slice(
    0,
    lastUsefulRevision.comments.indexOf('\n'),
  );
  return lastUsefulSummary;
};

// ensure all numbers display two digits
const formatNumber = (number: number): string => {
  if (number < 10) {
    return `0${number}`;
  } else return String(number);
};

const formatDate = (timestamp: number): string => {
  const date = new Date(timestamp * 1000);
  const day = date.getDate();
  const month = date.getMonth() + 1;
  const year = String(date.getFullYear()).slice(2);
  const hours = date.getHours();
  const minutes = date.getMinutes();

  return `${formatNumber(month)}/${formatNumber(day)}/${year} ${formatNumber(
    hours,
  )}:${formatNumber(minutes)}`;
};

const getTreeherderURL = (
  revision: Changeset['revision'],
  repository: Repository['name'],
) =>
  `https://treeherder.mozilla.org/jobs?repo=${repository}&revision=${revision}`;

const createDevtoolsDocsUrl = (
  supportedFramework: string,
  urlReadySuite: string,
  suite: string,
) => {
  let linkSupported = true;

  let devtoolsDocsURL = `${baseDocsURL}/devtools/tests/${supportedFramework}.html#${urlReadySuite}`;
  if (
    suite === removedOldTestDevTools ||
    nonDocumentedTestsDevTools.includes(suite)
  ) {
    devtoolsDocsURL = '';
    linkSupported = false;
  }
  return { devtoolsDocsURL, linkSupported };
};

const getDocsURL = (suite: string, framework_id: Framework['id']) => {
  const framework = frameworkMap[framework_id];
  const supportedFramework =
    supportedPerfdocsFrameworks[framework as SupportedPerfdocsFramework];
  const urlReadySuite = suite.replace(/:|\s|\.|_/g, '-').toLowerCase();

  let docsURL = '';
  let isLinkSupported = true;

  if (framework_id === devToolsFramework.id) {
    const { devtoolsDocsURL, linkSupported } = createDevtoolsDocsUrl(
      supportedFramework,
      urlReadySuite,
      suite,
    );

    isLinkSupported = linkSupported;
    docsURL = devtoolsDocsURL;
  } else if (supportedFramework) {
    docsURL = `${baseDocsURL}/testing/perfdocs/${supportedFramework}.html#${urlReadySuite}`;
  } else {
    isLinkSupported = false;
  }
  return { docsURL, isLinkSupported };
};

// TO DO: Review if this method is still needed
const setConfidenceClassName = (confidenceText: string | null) => {
  return confidenceText || 'unknown-confidence';
};

const swapArrayElements = <T>(
  array: T[],
  index1: number,
  index2: number,
): T[] => {
  if (
    index1 !== index2 &&
    [index1, index2].every((index) => index < array.length && index >= 0)
  ) {
    const newArray = [...array];
    [newArray[index1], newArray[index2]] = [newArray[index2], newArray[index1]];
    return newArray;
  }
  return array;
};

export {
  formatDate,
  getLatestCommitMessage,
  getTreeherderURL,
  setConfidenceClassName,
  swapArrayElements,
  truncateHash,
  getDocsURL,
};
