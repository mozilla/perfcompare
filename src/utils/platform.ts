import type { PlatformShortName } from '../types/state';

export const getPlatformShortName = (
  platformName: string,
): PlatformShortName => {
  if (platformName.toLowerCase().includes('linux')) return 'Linux';
  if (
    platformName.toLowerCase().includes('osx') ||
    platformName.toLowerCase().includes('os x')
  )
    return 'OSX';
  if (platformName.toLowerCase().includes('windows')) return 'Windows';
  if (platformName.toLowerCase().includes('android')) return 'Android';
  return 'Unspecified';
};
