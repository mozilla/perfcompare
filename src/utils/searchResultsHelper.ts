import type { Revision } from '../types/state';

const truncateHash = (revision: Revision['revision']) => revision.slice(0, 12);

// return only most recent commit message
// first commit is usually 'try_task_config'
const getLatestCommitMessage = (item: Revision) =>
  item.revisions[item.revisions.length - 1].comments;

export { truncateHash, getLatestCommitMessage };
