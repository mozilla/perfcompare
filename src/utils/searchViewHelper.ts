import type { Repository, Revision } from '../types/state';

const truncateHash = (revision: Revision['revision']) => revision.slice(0, 12);

// return only most recent commit message
// first commit is usually 'try_task_config'
const getLatestCommitMessage = (item: Revision) =>
  item.revisions[item.revisions.length - 1].comments;

// ensure all numbers display two digits
const formatNumber = (number: number): string => {
  if (number < 10) {
    return `0${number}`;
  } else return String(number);
};

const formatDate = (timestamp: number): string => {
  const date = new Date(timestamp * 1000);
  const day = date.getDate();
  const month = date.getMonth() + 1;
  const year = String(date.getFullYear()).slice(2);
  const hours = date.getHours();
  const minutes = date.getMinutes();

  return `${formatNumber(month)}/${formatNumber(day)}/${year} ${formatNumber(
    hours,
  )}:${formatNumber(minutes)}`;
};

const getTreeherderURL = (
  revision: Revision['revision'],
  repository: Repository['name'],
) =>
  `https://treeherder.mozilla.org/jobs?repo=${repository}&revision=${revision}`;

export { formatDate, getLatestCommitMessage, getTreeherderURL, truncateHash };
