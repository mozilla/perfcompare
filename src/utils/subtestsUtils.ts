import { subtestsOverTimeView } from '../common/constants';
import {
  getPerfherderSubtestsCompareWithBaseViewURL,
  getPerfherderSubtestsCompareOverTimeViewURL,
} from '../logic/treeherder';
import type {
  CompareResultsItem,
  SubtestsRevisionsHeader,
} from '../types/state';
import type { TimeRange } from '../types/types';

type SubtestsResults = {
  key: string;
  // By construction, there should be only one item in the array. But if more
  // than one subtests share the same name, then there will be more than one item.
  // Can this happen? We're not sure.
  value: CompareResultsItem[];
};

export function getSubtestsHeaderAndPerfherderURL(
  results: CompareResultsItem[],
  view: string,
  intervalValue: TimeRange['value'] | undefined,
): {
  subtestsHeader: SubtestsRevisionsHeader;
  subtestsViewPerfherderURL?: string;
} {
  const subtestsHeader: SubtestsRevisionsHeader = {
    suite: results[0].suite,
    framework_id: results[0].framework_id,
    test: results[0].test,
    option_name: results[0].option_name,
    extra_options: results[0].extra_options,
    new_rev: results[0].new_rev,
    new_repo: results[0].new_repository_name,
    base_rev: results[0].base_rev,
    base_repo: results[0].base_repository_name,
    base_parent_signature: results[0].base_parent_signature,
    new_parent_signature: results[0].base_parent_signature,
    platform: results[0].platform,
  };

  let subtestsViewPerfherderURL;
  if (
    subtestsHeader.base_parent_signature !== null &&
    subtestsHeader.new_parent_signature !== null
  ) {
    if (view === subtestsOverTimeView) {
      subtestsViewPerfherderURL = getPerfherderSubtestsCompareOverTimeViewURL(
        subtestsHeader.base_repo,
        subtestsHeader.new_repo,
        subtestsHeader.new_rev,
        subtestsHeader.framework_id,
        intervalValue ?? 86400,
        subtestsHeader.base_parent_signature,
        subtestsHeader.new_parent_signature,
      );
    } else
      subtestsViewPerfherderURL = getPerfherderSubtestsCompareWithBaseViewURL(
        subtestsHeader.base_repo,
        subtestsHeader.base_rev,
        subtestsHeader.base_repo,
        subtestsHeader.new_rev,
        subtestsHeader.framework_id,
        subtestsHeader.base_parent_signature,
        subtestsHeader.new_parent_signature,
      );
  }

  return { subtestsHeader, subtestsViewPerfherderURL };
}

const stringComparisonCollator = new Intl.Collator('en', {
  numeric: true,
  sensitivity: 'base',
});

export function defaultSortFunction(
  resultA: CompareResultsItem,
  resultB: CompareResultsItem,
) {
  return stringComparisonCollator.compare(resultA.test, resultB.test);
}

export function resultMatchesSearchTerm(
  result: CompareResultsItem,
  searchTerm: string,
) {
  return result.test.toLowerCase().includes(searchTerm.toLowerCase());
}

export function processResults(results: CompareResultsItem[]) {
  const processedResults = new Map<string, CompareResultsItem[]>();
  results.forEach((result) => {
    const { header_name: header } = result;
    const processedResult = processedResults.get(header);
    if (processedResult) {
      processedResult.push(result);
    } else {
      processedResults.set(header, [result]);
    }
  });
  const restructuredResults: SubtestsResults[] = Array.from(
    processedResults,
    function ([rowIdentifier, result]) {
      return {
        key: rowIdentifier,
        value: result,
      };
    },
  );

  return restructuredResults;
}
