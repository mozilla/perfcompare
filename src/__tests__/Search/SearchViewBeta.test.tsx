import { renderHook } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

import SearchViewBeta from '../../components/Search/beta/SearchView';
import useProtocolTheme from '../../theme/protocolTheme';
import { renderWithRouter } from '../utils/setupTests';

describe('Search View', () => {
  const protocolTheme = renderHook(() => useProtocolTheme()).result.current
    .protocolTheme;

  const toggleColorMode = renderHook(() => useProtocolTheme()).result.current
    .toggleColorMode;

  it('renders correctly when there are no results', async () => {
    renderWithRouter(
      <SearchViewBeta
        toggleColorMode={toggleColorMode}
        protocolTheme={protocolTheme}
      />,
    );

    expect(document.body).toMatchSnapshot();
    await act(async () => void jest.runOnlyPendingTimers());
  });
});
