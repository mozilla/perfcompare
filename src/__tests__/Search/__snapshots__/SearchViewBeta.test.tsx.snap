// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search View renders correctly when there are no results 1`] = `
<body>
  <div>
    <section
      class="perfcompare-body"
    >
      <div
        class="header-container f7qotpz"
      >
        <div
          class="toggle-dark-mode f1vchea6 MuiBox-root css-0"
        >
          <div
            class="MuiTypography-root MuiTypography-body1 MuiTypography-alignLeft toggle-text css-7idqxi-MuiTypography-root"
          >
            Dark mode
          </div>
          <span
            class="MuiSwitch-root MuiSwitch-sizeMedium toggle-switch css-julti5-MuiSwitch-root"
          >
            <span
              class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
            >
              <input
                aria-label="Dark mode switch"
                class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                type="checkbox"
              />
              <span
                class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
            />
          </span>
        </div>
        <div
          class="header-text MuiBox-root css-0"
        >
          <div
            class="MuiTypography-root MuiTypography-h1 MuiTypography-alignCenter MuiTypography-gutterBottom perfcompare-header css-8mdcpf-MuiTypography-root"
          >
            PerfCompare
          </div>
          <div
            class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter MuiTypography-gutterBottom perfcompare-tagline css-1crpcdz-MuiTypography-root"
          >
            Analyze results of performance tests to detect regressions and identify opportunities for improvement.
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium learn-more-btn css-13bs7uq-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Learn more
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </section>
  </div>
</body>
`;
